File: src/app/repository/CommentRepository.java
package app.repository;

import app.model.Comment;

import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.util.ArrayList;
import java.util.List;

public final class CommentRepository extends _BaseRepository {
  private CommentRepository() {
  }

  public static Comment getComment(int commentId) {
    final String SQL = "SELECT * from `CommentCounted` where `commentId`=?;";
    return connect(connection -> {
      final PreparedStatement statement = connection.prepareStatement(SQL);
      statement.setInt(1, commentId);
      final ResultSet res = statement.executeQuery();
      return res.next() ? Comment.from(res) : null;
    });
  }

  /**
   * returns all comment and replies on this articleId
   */
  public static List<Comment> listComments(int articleId) {
    final String SQL = "SELECT * from `CommentCounted` where `articleId`=?; --and `reply_commentId` is null";
    return connect(connection -> {
      final PreparedStatement statement = connection.prepareStatement(SQL);
      statement.setInt(1, articleId);
      final ResultSet res = statement.executeQuery();
      final List<Comment> result = new ArrayList<>();
      while (res.next())
        result.add(Comment.from(res));
      return result;
    });
  }

  public static Comment deleteComment(int commentId) {
    final String SQL = "DELETE FROM `Comment` WHERE `commentId`=? RETURNING *;";
    return connect(connection -> {
      final PreparedStatement statement = connection.prepareStatement(SQL);
      statement.setInt(1, commentId);
      final ResultSet res = statement.executeQuery();
      return res.next() ? Comment.from(res) : null;
    });
  }

  /**
   * add comment to an {@link app.model.Article} or reply on another {@link app.model.Comment} (if replyCommentId>0)
   */
  public static Comment commentOn(int userId, int articleId, int replyCommentId, String content) {
    final String SQL_C = "INSERT INTO `Comment` (userId, content, time, notified, articleId, reply_commentId) VALUES (?,?,?,0,?,null) RETURNING *;";
    final String SQL_R = "INSERT INTO `Comment` (userId, content, time, notified, articleId, reply_commentId) SELECT ?,?,?,0,articleId,commentId from Comment where commentId=? RETURNING *;";
    return connect(connection -> {
      final PreparedStatement statement = connection.prepareStatement(replyCommentId > 0 ? SQL_R : SQL_C);
      statement.setInt(1, userId);
      statement.setString(2, content);
      statement.setLong(3, System.currentTimeMillis());
      statement.setInt(4, replyCommentId > 0 ? replyCommentId : articleId);
      final ResultSet res = statement.executeQuery();
      return res.next() ? Comment.from(res) : null;
    });
  }

}

File: src/app/repository/CommentRepository.kt
@file:JvmName("Repository")
@file:JvmMultifileClass

package app.repository

import app.model.Comment

fun getCommentById(commentId: Int): Comment? =
	connect {
		val SQL = """SELECT * from `CommentCounted` where `commentId`=?;"""
		val statement = it.prepareStatement(SQL)
		statement.setInt(1, commentId)
		statement.executeQuery()
			.singleOf<Comment>()
	}

/**
 * returns all comment and replies on this articleId
 */
fun listComments(articleId: Int): List<Comment> =
	connect {
		val SQL = """SELECT * from `CommentCounted` where `articleId`=?; --and `reply_commentId` is null"""
		val statement = it.prepareStatement(SQL)
		statement.setInt(1, articleId)
		statement.executeQuery()
			.listOf<Comment>()
	} ?: emptyList()

fun deleteComment(commentId: Int): Comment? =
	connect {
		val SQL = """DELETE FROM `Comment` WHERE `commentId`=? RETURNING *;"""
		val statement = it.prepareStatement(SQL)
		statement.setInt(1, commentId)
		statement.executeQuery()
			.singleOf<Comment>()
	}

/**
 * add comment to an [app.model.Article] or reply on another [app.model.Comment] (if replyCommentId>0)
 */
fun sendCommentOn(userId: Int, articleId: Int, replyCommentId: Int, content: String): Comment? =
	connect {
		val SQL_C = """INSERT INTO `Comment`
			(`userId`, `content`, `time`, `notified`, `articleId`, `reply_commentId`)
			VALUES (?,?,?,0,?,null) RETURNING *;"""
		val SQL_R = """INSERT INTO `Comment`
			(`userId`, `content`, `time`, `notified`, `articleId`, `reply_commentId`)
			SELECT ?,?,?,0,`articleId`,`commentId` from `Comment` where `commentId`=? RETURNING *;"""
		val statement = if (replyCommentId > 0)
			it.prepareStatement(SQL_R)
		else it.prepareStatement(SQL_C)
		statement.setInt(1, userId)
		statement.setString(2, content)
		statement.setLong(3, System.currentTimeMillis())
		statement.setInt(4, if (replyCommentId > 0) replyCommentId else articleId)
		statement.executeQuery()
			.singleOf<Comment>()
	}
